src/NodeHandler.ts

```diff
@@ -0,0 +1,44 @@
+import { Evented } from '@dojo/core/Evented';
+import { VNodeProperties } from '@dojo/interfaces/vdom';
+import Map from '@dojo/shim/Map';
+
+export enum Type {
+	Projector = 'Projector',
+	Widget = 'Widget'
+}
+
+export default class NodeHandler extends Evented {
+
+	private _nodeMap = new Map<string, Element>();
+
+	public get(key: string): Element | undefined {
+		return this._nodeMap.get(key);
+	}
+
+	public has(key: string): Boolean {
+		return this._nodeMap.has(key);
+	}
+
+	public add(element: Element, properties: VNodeProperties) {
+		const key = String(properties.key);
+		this._nodeMap.set(key, element);
+		this.emit({ type: key });
+	}
+
+	public addRoot(element: Element, properties: VNodeProperties) {
+		this.add(element, properties);
```

Comment:
Do you need the key check for adding the root too?

Reply:
good catch

Reply:
can we add a failing test for this if it needs the guard 😄

Reply:
way ahead of you 😉